src/daemonizer/posix_fork.cpp

```diff
@@ -21,15 +21,42 @@
 namespace posix {
 
 namespace {
-  void quit(std::string const & message)
+  void quit(const std::string & message)
   {
     LOG_ERROR(message);
     throw std::runtime_error(message);
   }
 }
 
-void fork()
+void fork(const std::string & pidfile)
 {
+  // If a PID file is specified, we open the file here, because
+  // we can't report errors after the fork operation.
+  // When we fork, we close thise file in each of the parent
+  // processes.
+  // Only in the final child process do we write the PID to the
+  // file (and close it).
+  std::ofstream pidofs;
+  if (! pidfile.empty ())
+  {
+	int oldpid;
+    std::ifstream pidrifs;
+    pidrifs.open(pidfile, std::fstream::in);
+    if (! pidrifs.fail())
+    {
+	  if (pidrifs >> oldpid && oldpid > 1 && kill(oldpid, 0) == 0)
+      {
+        quit("PID file " + pidfile + " already exists and the PID therein is valid");
```

Comment:
Well aware it's optional! Just saying the only reason the daemon should not be allowed to run is a catastrophic reason, not just because there's an existing pid file. Other code deals with those catastrophic cases so your pidfile code does not need to make any assumptions.